File: src/main/java/com/kasisoft/libs/common/functional/KSupplier.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KSupplier<T> {

    T get() throws Exception;

    @NotNull
    default Supplier<T> protect() {
        return () -> {
            try {
                return get();
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KSupplier.kt
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import java.util.function.*;

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KSupplier<R: Any?> {

    @Throws(Exception::class)
    fun get(): R

    fun protect(): Supplier<R> =
        Supplier<R> {
            try {
                get()
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
